tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/PoPTests.cs

```diff
@@ -0,0 +1,185 @@
+                 protectedResournceUri1,
+                 HttpMethod.Get).ConfigureAwait(false);
+
+            // recreate the pca to ensure that the silent call is served from the cache, i.e. the key remains stable
+            pca = PublicClientApplicationBuilder
+                .Create(clientId)
+                .WithHttpClientFactory(new NoAccessHttpClientFactory()) // token should be served from the cache, no network access necessary
+                .Build();
+            ConfigureInMemoryCache(pca);
+
+            var accounts = await pca.GetAccountsAsync().ConfigureAwait(false);
+            authResult = await pca
+                .AcquireTokenSilent(s_scopes, accounts.Single())
+                .WithPoPAuthenticationScheme(protectedResournceUri1, HttpMethod.Get)
+                .ExecuteAsync()
+                .ConfigureAwait(false);
+
+            await VerifyPoPTokenAsync(
+                clientId, 
+                authResult.CreateAuthorizationHeader(),
+                protectedResournceUri1,
+                HttpMethod.Get).ConfigureAwait(false);
+
+            // Call some other Uri - the same pop assertion can be reused, i.e. no need to call Evo
+            authResult = await pca
+              .AcquireTokenSilent(s_scopes, accounts.Single())
+              .WithPoPAuthenticationScheme(protectedResournceUri2, HttpMethod.Post)
+              .ExecuteAsync()
+              .ConfigureAwait(false);
+
+            await VerifyPoPTokenAsync(
+                clientId,
+                authResult.CreateAuthorizationHeader(),
+                protectedResournceUri2,
+                HttpMethod.Post).ConfigureAwait(false);
+        }
+
+        private class NoAccessHttpClientFactory : IMsalHttpClientFactory
+        {
+            private const string Message = "Not expecting to make HTTP requests.";
+
+            public HttpClient GetHttpClient()
+            {
+                Assert.Fail(Message);
+                throw new InvalidOperationException(Message);
+            }
+        }
+
+        /// <summary>
+        /// This calls a special endpoint that validates any POP token against a configurable HTTP request.
+        /// The HTTP request is configured through headers.
+        /// </summary>
+        private async Task VerifyPoPTokenAsync(string clientId, string popAuthHeader, Uri uri, HttpMethod httpMethod)
+        {
+            var httpClient = new HttpClient();
+            HttpResponseMessage response;
+            var request = new HttpRequestMessage(HttpMethod.Post, PoPValidatorEndpoint);
+
+            request.Headers.Add("Authorization", popAuthHeader);
+            request.Headers.Add("Secret", _popValidationEndpointSecret);
+            request.Headers.Add("Authority", "https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/");
+            request.Headers.Add("ClientId", clientId);
+
+            // the URI the POP token is bound to
+            request.Headers.Add("ShrUri", uri.ToString());
```

Comment:
>Shr [](start = 33, length = 3)

what's shr?

Reply:
George has setup a small web api endpoint where we pass an HttpRequest with a POP token plus some headers and it checks that the POP token is according to those headers. So in this case, it will check that

`popToken["bound_uri"] == headers["ShrUri"]`

Reply:
(otherwise I also don't know what `shr` stands for :))

Reply:
it's SignedHttpRequest...I asked George :)

---
In reply to: [352711137](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1525#discussion_r352711137) [](ancestors = 352711137)

Reply:
I see...thanks

---
In reply to: [352710544](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1525#discussion_r352710544) [](ancestors = 352710544)